src/cryptonote_core/blockchain.cpp

```diff
@@ -1303,7 +1349,11 @@ difficulty_type Blockchain::get_next_difficulty_for_alternative_chain(const std:
   // based on its blocks alone, need to get more blocks from the main chain
   if(alt_chain.size()< DIFFICULTY_BLOCKS_COUNT)
   {
-    CRITICAL_REGION_LOCAL(m_blockchain_lock);
+    m_blockchain_transaction.start_write();
+    epee::misc_utils::auto_scope_leave_caller scope_exit_handler = 
+      epee::misc_utils::create_scope_leave_handler([&](){
+        m_blockchain_transaction.end_write();
+      });
```

Comment:
I feel like this pattern is a good candidate for an [RAII class](https://en.cppreference.com/w/cpp/language/raii). This class could also let you lock and unlock more finely within this scope depending on the operations required.

Reply:
I disagree here. Eventually, we want to have a fine-grained locking mechanism. RAII here would just make this a lot more cumbersome to use. 

Consider this as prototype/experimental PR, which does have bugs, that is why I am using `auto_scope_leave_caller` here. Eventually, we want to only lock where we interact with write/reads. Not whole functions.

Reply:
You can still have fine-grained locking with RAII. Consider [`std::unique_lock`](https://en.cppreference.com/w/cpp/thread/unique_lock).